### PR TITLE
admindcli helper and test framework improvement (#9656)

### PR DESCRIPTION
**Pull request by andy-fong**
# Description

While working on the test for https://github.com/solo-io/solo-projects/issues/5455, found an issue with GetSingleListenerFromDynamicListeners() in admincli where it would crash if the config_dump returns an empty array. 

Also, when running focused tests, the docker containers is left behind. 

## Code changes

- return error in GetSingleListenerFromDynamicListeners() if the config array is empty
- check docker container no longer exists before finishing AfterEach()

## Docs changes

# Context

## Notes for reviewers

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: pkg/utils/envoyutils/admincli/client.go [hunk 1]
```diff
@@ -239,8 +239,13 @@ func (c *Client) GetSingleListenerFromDynamicListeners(
 		return nil, fmt.Errorf("could not get envoy config_dump from adminClient: %w", err)
 	}
 
+	configs := cfgDump.GetConfigs()
+	if len(configs) == 0 {
+		return nil, fmt.Errorf("could not get config: config is empty")
```
Comment (nfuden) [inline pkg/utils/envoyutils/admincli/client.go:244]: can we add a comment as to where we can get a valid configdump but have no configdump?

File: test/e2e/test_context.go [hunk 2]
```diff
@@ -141,6 +141,9 @@ func (c *TestContext) AfterEach() {
 	ginkgo.By("TestContext.AfterEach: Cancelling test context")
 	// All services connected to the TestContext are tied to the context, so cancelling it will clean those up
 	c.cancel()
+	// This line prevents the process to quit before the envoy docker process is killed when running
+	// Focused Tests
```
Comment (nfuden) [inline test/e2e/test_context.go:145]: how does this work with the environment variable for no tear own? can we show how and why this is safe?

**Review by sam-heilbron — APPROVED** [review]

File: pkg/utils/envoyutils/admincli/client.go (hunk 1 above)
Comment (andy-fong) [reply pkg/utils/envoyutils/admincli/client.go:244]: This is the case where I dump the config too soon and there is no dynamic_listener setup yet. Also, with the regex match, if no listener matches, it returns empy object: 
```
Solo-System-AFong:solo-projects andyfong$ curl -v "localhost:20001/config_dump?resource=dynamic_listeners&name_regex=foo"
* Host localhost:20001 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:20001...
* Connected to localhost (::1) port 20001
> GET /config_dump?resource=dynamic_listeners&name_regex=foo HTTP/1.1
> Host: localhost:20001
> User-Agent: curl/8.9.1
> Accept: */*
> 
* Request completely sent off
< HTTP/1.1 200 OK
< content-type: application/json
< cache-control: no-cache, max-age=0
< x-content-type-options: nosniff
< date: Tue, 17 Sep 2024 00:20:48 GMT
< server: envoy
< transfer-encoding: chunked
< 
{}```

I will add a comment for both case.

File: test/e2e/test_context.go (hunk 2 above)
Comment (andy-fong) [reply test/e2e/test_context.go:145]: @nfuden I was not aware of a TEAR_DOWN env. So, thank you for bringing that up. Are you talking about this: https://github.com/solo-io/gloo/blob/bddae6d73f635c55fee089b36fd9db7b3be9fdda/test/testutils/env.go#L10 ?
Looks like that only get used for kube2e test to tear down Gloo in k8s. In e2e test, gloo is in memory or inside the test process itself (I think).

**Review by nfuden — APPROVED** [review]

**Pull request merged**